program_transformers/src/rollups/rollup_persister.rs

```diff
@@ -0,0 +1,571 @@
+        Self {
+            instruction: InstructionName::MintV1,
+            tree_update: Some((&value.tree_update).into()),
+            leaf_update: Some(LeafSchemaEvent::new(
+                Version::V1,
+                value.leaf_update.clone(),
+                hash,
+            )),
+            payload: Some(Payload::MintV1 {
+                args: value.mint_args.clone(),
+                authority: value.authority,
+                tree_id: value.tree_update.id,
+            }),
+        }
+    }
+}
+
+#[automock]
+#[async_trait]
+pub trait RollupDownloader {
+    async fn download_rollup(&self, url: &str) -> Result<Box<Rollup>, RollupValidationError>;
+    async fn download_rollup_and_check_checksum(
+        &self,
+        url: &str,
+        checksum: &str,
+    ) -> Result<Box<Rollup>, RollupValidationError>;
+}
+
+pub struct RollupPersister<T: ConnectionTrait + TransactionTrait, D: RollupDownloader> {
+    txn: Arc<T>,
+    downloader: D,
+    cl_audits: bool,
+}
+
+pub struct RollupDownloaderForPersister {}
+
+#[async_trait]
+impl RollupDownloader for RollupDownloaderForPersister {
+    async fn download_rollup(&self, url: &str) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+
+    async fn download_rollup_and_check_checksum(
+        &self,
+        url: &str,
+        checksum: &str,
+    ) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        let file_hash = xxhash_rust::xxh3::xxh3_128(&response);
+        let hash_hex = hex::encode(file_hash.to_be_bytes());
+        if hash_hex != checksum {
+            return Err(RollupValidationError::InvalidDataHash(
+                checksum.to_string(),
+                hash_hex,
+            ));
+        }
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+}
+
+impl<T: ConnectionTrait + TransactionTrait, D: RollupDownloader> RollupPersister<T, D> {
+    pub fn new(txn: Arc<T>, downloader: D, cl_audits: bool) -> Self {
+        Self {
+            txn,
+            downloader,
+            cl_audits,
+        }
+    }
+
+    pub async fn persist_rollups(&self) {
+        loop {
+            tokio::time::sleep(Duration::from_secs(5)).await;
+            let Ok((rollup_to_verify, rollup)) = self.get_rollup_to_verify().await else {
+                continue;
+            };
+            let Some(rollup_to_verify) = rollup_to_verify else {
+                // no rollups to persist
+                continue;
+            };
+            let Ok(rollup) = rollup
+                .map(|r| bincode::deserialize::<Rollup>(r.rollup_binary_bincode.as_slice()))
+                .transpose()
+                .map_err(|e| ProgramTransformerError::DeserializationError(e.to_string()))
+            else {
+                continue;
+            };
```

Comment:
In the case of error will it keep trying to deserialize this rollup in subsequent iterations of the loop, because it is stuck in `RollupPersistingState::ReceivedTransaction` and never deleted from the table?

Reply:
Great catch, thank you!

Reply:
I think the sequence would be: if this deserialization failed, it would default to `None`, which it would pass to `persist_rollup`, which would re-download and serialize it, and update the row in the `rollup` table with new `FileHash` and `RollupBinaryBincode`, which we would assume would fix the issue.  Is this correct understanding?

But now looking at it more, when is this code even expected to run?  In the outer `persist_rollups` process, it should get a rollup to verify from the `rollup_to_verify` table and at the same time attempt to get the downloaded rollup from the `rollup` table.  But it doesn't seem like the Rollup should yet be poplulated in the rollup table until `persist_rollup` function is called which actually does the download.